### PR TITLE
Fix BUG IN FGRAPH.REPLACE OR A LISTENER

### PR DESCRIPTION
**Pull request by nouiz**
Don't merge now, but it can be reviewed.

I need to check if there isn't extra scan in the graph. I'm not sure about that.

fix gh-3663

### CODE REVIEW

**Comment by carriepl** [comment]
I'm not familiar enough with this part of theano to review. @abergeron or @lamblin, can you have a look?

**Comment by nouiz** [comment]
I did some check. This PR fix the crash, but will have 4 scan in my test
case instead of 2 at commit 146fbae8c4

@carriepl can you check why while the tests run with your modif to
DebugMode? Can you make a PR with your DebugMode change?

**Comment by carriepl** [comment]
My changes to debugmode are in PR #3713. It might not be 100% ready to merge but it what I'm currently running to see the ops that would need changing to remove the checks in Scan.

**Comment by nouiz** [comment]
I updated this PR.

**Comment by lamblin** [comment]
OK to merge if Travis passes.

File: theano/scan_module/scan_op.py [hunk 1]
```diff
@@ -1607,14 +1607,18 @@ def infer_shape(self, node, input_shapes):
...
         # of iterations.
         # sequences
-        seqs_shape = [x[1:] for x in input_shapes[1:1 + self.n_seqs]]
-        inner_seqs = self.inputs[:self.n_seqs]
-        outer_seqs = node.inputs[1:1 + self.n_seqs]
-        for in_s, out_s in izip(inner_seqs, outer_seqs):
-            out_equivalent[in_s] = out_s[0]
+        # seqs_shape = [x[1:] for x in input_shapes[1:1 + self.n_seqs]]
+        # inner_seqs = self.inputs[:self.n_seqs]
+        # outer_seqs = node.inputs[1:1 + self.n_seqs]
```
Comment (lamblin) [inline theano/scan_module/scan_op.py]: These variables are still used below, so maybe there is more to comment out, or we should not comment out all of that.

**Comment by nouiz** [comment]
Thanks for the comment. I updated this PR.

**Comment by nouiz** [comment]
I amended the last commit.

**Comment by nouiz** [comment]
I updated the tests in the PR to match the new behavior.

This trigger what I think is another not related bug in Scan that generate a crash. So I commented the execution of the problematic case. I think we should merge this PR as it leave the master in a better state then the current state of Theano and make an issue for the crash.

@carriepl you talked about a bug in scan that need fixing while working on the problematic PR. Was it this case?

**Comment by nouiz** [comment]
IRL, @carriepl told me we can merge this PR. The failure of the new tests was there before and if we return not just the shape, but the real data, it crash. I'll create another issue for that.

**Pull request merged**